File: Counting-Sort/python/leetcode_lc_912.py
nums = [1,1,5,0,2,0]

print(nums)

class Solution:
    def sortArray(self, nums):
        
        for i in range(len(nums)):
            if i == 0:
                max = min = nums[i]
            if nums[i] > max:
                max = nums[i]
            if nums[i] < min:
                min = nums[i]

        if min < 0:   
            nums = [v-min for v in nums]
            max -= min
        else:
            min = 0

        return self.countingSort(nums,max,min)

    def countingSort(self,arr,max,min):
        bucket = [0]*(max+1)
        idx = 0
        for i in range(len(arr)):
            if not bucket[arr[i]]:
                bucket[arr[i]] = 1
            else:
                bucket[arr[i]] += 1

        for j in range(max+1):
            while bucket[j]:
                arr[idx] = j + min
                idx += 1
                bucket[j] -= 1

        return arr


solution = Solution()

print("升序{}".format(nums,solution.sortArray(nums)))

File: Counting-Sort/js/leetcode_lc_912.js
const arr = []

function initData(count){
    for(let i = 0 ; i < count ; i++){
        arr.push(Math.floor(Math.random()*50-20))
    }
}

initData(6)

console.log(arr)

var sortArray = function(nums) {
    
    let min = max = 0
    for(let i = 0 ; i < nums.length ; i++){
        if(i === 0) min = max = nums[i]

        if(nums[i] > max) max = nums[i]

        if(nums[i] < min) min = nums[i]
    }

    if(min < 0){
        nums = nums.map(v => v-min)
        max -= min
    }else{
        min = 0
    }

    return countingSort(nums,max,min)
};


function countingSort(arr,max,min){
    // console.log(max,min)
    const bucket = new Array(max+1)
    let sortedIdx = 0

    for(let i = 0 ; i < arr.length ; i++){
        if(!bucket[arr[i]]){
            bucket[arr[i]] = 1
        }else{
            bucket[arr[i]]++
        }
    }

    for(let j = 0 ; j < bucket.length ; j++){
        while(bucket[j] > 0){
            arr[sortedIdx] = j + min
            sortedIdx++
            bucket[j]--
        }
    }
    return arr
}



console.log(sortArray(arr))
